File: backend/correlation/scorer.py
"""Explainable, deterministic V0 correlation scoring."""

from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterable, Mapping

from .features import (
    Classification,
    Direction,
    EvidenceContribution,
    EvidenceFamily,
    EvidenceSignal,
    PairAssessment,
    SignalType,
)

DEFAULT_WEIGHTS: Mapping[SignalType, float] = {
    SignalType.EMAIL_EXACT: 45.0,
    SignalType.DIRECT_PROFILE_LINK: 40.0,
    SignalType.SHARED_PERSONAL_DOMAIN: 30.0,
    SignalType.DOMAIN_EXACT: 30.0,
    SignalType.URL_EXACT: 25.0,
    SignalType.SHARED_IDENTIFIER: 25.0,
    SignalType.USERNAME_EXACT: 20.0,
    SignalType.REPOSITORY_RELATIONSHIP: 20.0,
    SignalType.USERNAME_SIMILARITY: 12.0,
    SignalType.DISPLAY_NAME_SIMILARITY: 10.0,
    SignalType.PROJECT_OVERLAP: 15.0,
    SignalType.ORGANIZATION_OVERLAP: 10.0,
    SignalType.BIO_SIMILARITY: 8.0,
    SignalType.TOPIC_OVERLAP: 3.0,
    SignalType.LOCATION_OVERLAP: 5.0,
    SignalType.EXACT_AVATAR: 20.0,
    SignalType.IMAGE_SIMILARITY: 16.0,
    SignalType.FACE_SIMILARITY: 16.0,
    SignalType.DIFFERENT_FACE: 40.0,
    SignalType.EXPLICIT_IDENTITY_CONFLICT: 40.0,
    SignalType.PERSONAL_DOMAIN_CONFLICT: 20.0,
    SignalType.BIOGRAPHICAL_CONTRADICTION: 15.0,
    SignalType.LOCATION_CONFLICT: 15.0,
    SignalType.ORGANIZATION_CONFLICT: 15.0,
}

_ANCHOR_SIGNALS = {
    SignalType.EMAIL_EXACT,
    SignalType.DIRECT_PROFILE_LINK,
    SignalType.SHARED_PERSONAL_DOMAIN,
    SignalType.DOMAIN_EXACT,
    SignalType.EXACT_AVATAR,
    SignalType.PROJECT_OVERLAP,
}


class CorrelationScorer:
    """Score evidence while counting at most one signal per evidence family."""

    def __init__(
        self,
        weights: Mapping[SignalType, float] | None = None,
        *,
        model_version: str = "deterministic-v0.1",
    ) -> None:
        self.weights = dict(DEFAULT_WEIGHTS if weights is None else weights)
        self.model_version = model_version

    def score(
        self,
        evidence: Iterable[EvidenceSignal],
        *,
        left_profile_id: str | None = None,
        right_profile_id: str | None = None,
    ) -> PairAssessment:
        all_evidence = tuple(evidence)
        if all_evidence:
            pair_keys = {item.pair_key for item in all_evidence}
            if len(pair_keys) != 1:
                raise ValueError("all evidence must describe one profile pair")
            inferred_left, inferred_right = next(iter(pair_keys))
            left_profile_id = left_profile_id or inferred_left
            right_profile_id = right_profile_id or inferred_right
            if {left_profile_id, right_profile_id} != {inferred_left, inferred_right}:
                raise ValueError("explicit profile IDs do not match evidence")
        elif not left_profile_id or not right_profile_id:
            raise ValueError("profile IDs are required when scoring empty evidence")
        if left_profile_id == right_profile_id:
            raise ValueError("cannot score a profile against itself")

        selected = self._select_by_family(all_evidence)
        contributions = tuple(
            EvidenceContribution(item, self.contribution(item)) for item in selected
        )
        raw_score = round(sum(item.weighted_score for item in contributions), 6)
        normalized_score = round(max(-1.0, min(1.0, raw_score / 100.0)), 6)

        support_families = {
            item.evidence_family for item in selected if item.direction is Direction.SUPPORT
        }
        contradiction_families = {
            item.evidence_family for item in selected if item.direction is Direction.CONTRADICT
        }
        classification = self._classify(
            raw_score,
            selected,
            len(support_families),
        )

        # Track independent sources contributing to correlation
        independent_sources = set()
        matching_idents = []
        for item in selected:
            if item.direction is Direction.SUPPORT:
                src = item.source_key or item.metadata.get("source") or item.evidence_family.value
                independent_sources.add(src)
                ident_val = item.metadata.get("identifier_value") or item.metadata.get("unusualness")
                if ident_val:
                    matching_idents.append(str(ident_val))

        independent_source_count = max(1, len(independent_sources)) if support_families else 0

        ordered_left, ordered_right = sorted((left_profile_id, right_profile_id))
        return PairAssessment(
            left_profile_id=ordered_left,
            right_profile_id=ordered_right,
            raw_score=raw_score,
            normalized_score=normalized_score,
            classification=classification,
            evidence=all_evidence,
            selected_evidence=selected,
            contributions=contributions,
            support_family_count=len(support_families),
            contradiction_family_count=len(contradiction_families),
            model_version=self.model_version,
            independent_source_count=independent_source_count,
            matching_identifiers=tuple(dict.fromkeys(matching_idents)),
        )

    def contribution(self, item: EvidenceSignal) -> float:
        weight = self.weights.get(item.signal_type, 0.0)
        sign = -1.0 if item.direction is Direction.CONTRADICT else 1.0
        if item.direction is Direction.NEUTRAL:
            sign = 0.0
        return round(sign * weight * item.normalized_score * item.reliability, 6)

    def _select_by_family(
        self,
        evidence: tuple[EvidenceSignal, ...],
    ) -> tuple[EvidenceSignal, ...]:
        grouped: dict[EvidenceFamily, list[EvidenceSignal]] = defaultdict(list)
        for item in evidence:
            if item.direction is not Direction.NEUTRAL:
                grouped[item.evidence_family].append(item)

        selected: list[EvidenceSignal] = []
        for family in sorted(grouped, key=str):
            # Correlated signals (exact hash, pHash, and face similarity, for
            # example) count once. Equal-strength ties prefer contradiction to
            # avoid optimistic identity merges.
            strongest = max(
                grouped[family],
                key=lambda item: (
                    abs(self.contribution(item)),
                    item.direction is Direction.CONTRADICT,
                    item.signal_type.value,
                ),
            )
            selected.append(strongest)
        return tuple(selected)

    def _classify(
        self,
        raw_score: float,
        selected: tuple[EvidenceSignal, ...],
        support_family_count: int,
    ) -> Classification:
        if raw_score <= -10.0:
            return Classification.CONTRADICTORY

        has_anchor = any(
            item.direction is Direction.SUPPORT and item.signal_type in _ANCHOR_SIGNALS
            for item in selected
        )
        has_direct_link = any(
            item.direction is Direction.SUPPORT
            and item.signal_type is SignalType.DIRECT_PROFILE_LINK
            and self.contribution(item) >= 35.0
            for item in selected
        )
        material_contradiction = any(
            item.direction is Direction.CONTRADICT and abs(self.contribution(item)) >= 15.0
            for item in selected
        )

        if (
            raw_score >= 40.0
            and support_family_count >= 2
            and has_anchor
            and not material_contradiction
        ):
            return Classification.STRONG
        if (
            raw_score >= 20.0
            and not material_contradiction
            and (support_family_count >= 2 or has_direct_link)
        ):
            return Classification.LIKELY
        if raw_score >= 7.0:
            return Classification.AMBIGUOUS
        return Classification.WEAK


def score_evidence(
    evidence: Iterable[EvidenceSignal],
    *,
    left_profile_id: str | None = None,
    right_profile_id: str | None = None,
) -> PairAssessment:
    """Score with the default, versioned deterministic model."""

    return CorrelationScorer().score(
        evidence,
        left_profile_id=left_profile_id,
        right_profile_id=right_profile_id,
    )

File: src/correlation/scorer.rs
use crate::correlation::features::{Direction, EvidenceFamily, EvidenceSignal, SignalType};
use crate::db::models::Classification;
use std::collections::{HashMap, HashSet};

pub fn get_signal_weight(signal_type: SignalType) -> f64 {
    match signal_type {
        SignalType::EmailExact => 45.0,
        SignalType::DirectProfileLink => 40.0,
        SignalType::SharedPersonalDomain => 30.0,
        SignalType::DomainExact => 30.0,
        SignalType::UrlExact => 25.0,
        SignalType::SharedIdentifier => 25.0,
        SignalType::UsernameExact => 20.0,
        SignalType::RepositoryRelationship => 20.0,
        SignalType::UsernameSimilarity => 12.0,
        SignalType::DisplayNameSimilarity => 10.0,
        SignalType::ProjectOverlap => 15.0,
        SignalType::OrganizationOverlap => 10.0,
        SignalType::BioSimilarity => 8.0,
        SignalType::TopicOverlap => 3.0,
        SignalType::LocationOverlap => 5.0,
        SignalType::ExactAvatar => 20.0,
        SignalType::ImageSimilarity => 16.0,
        SignalType::FaceSimilarity => 16.0,
        SignalType::DifferentFace => 40.0,
        SignalType::ExplicitIdentityConflict => 40.0,
        SignalType::PersonalDomainConflict => 20.0,
        SignalType::BiographicalContradiction => 15.0,
        SignalType::LocationConflict => 15.0,
        SignalType::OrganizationConflict => 15.0,
        SignalType::NameSearchMention => 4.0,
    }
}

pub fn calculate_contribution(signal: &EvidenceSignal) -> f64 {
    let weight = get_signal_weight(signal.signal_type);
    let sign = match signal.direction {
        Direction::Support => 1.0,
        Direction::Contradict => -1.0,
        Direction::Neutral => 0.0,
    };
    sign * weight * signal.normalized_score * signal.reliability
}

#[derive(Debug, Clone)]
pub struct ScoredPair {
    pub raw_score: f64,
    pub normalized_score: f64,
    pub classification: Classification,
    pub support_family_count: usize,
    pub contradiction_family_count: usize,
    pub selected_evidence: Vec<EvidenceSignal>,
}

pub fn score_evidence(signals: &[EvidenceSignal]) -> ScoredPair {
    // Group signals by EvidenceFamily and select strongest contribution
    let mut family_groups: HashMap<EvidenceFamily, Vec<&EvidenceSignal>> = HashMap::new();
    for sig in signals {
        if sig.direction != Direction::Neutral {
            family_groups.entry(sig.evidence_family).or_default().push(sig);
        }
    }

    let mut selected: Vec<EvidenceSignal> = Vec::new();
    for (_family, sigs) in family_groups {
        if let Some(&best) = sigs.iter().max_by(|a, b| {
            let contrib_a = calculate_contribution(a).abs();
            let contrib_b = calculate_contribution(b).abs();
            contrib_a.partial_cmp(&contrib_b).unwrap_or(std::cmp::Ordering::Equal)
        }) {
            selected.push(best.clone());
        }
    }

    let raw_score: f64 = selected.iter().map(calculate_contribution).sum();
    let normalized_score = (raw_score / 100.0).clamp(-1.0, 1.0);

    let support_families: HashSet<_> = selected
        .iter()
        .filter(|s| s.direction == Direction::Support)
        .map(|s| s.evidence_family)
        .collect();

    let contradiction_families: HashSet<_> = selected
        .iter()
        .filter(|s| s.direction == Direction::Contradict)
        .map(|s| s.evidence_family)
        .collect();

    let classification = classify(raw_score, &selected, support_families.len());

    ScoredPair {
        raw_score,
        normalized_score,
        classification,
        support_family_count: support_families.len(),
        contradiction_family_count: contradiction_families.len(),
        selected_evidence: selected,
    }
}

fn classify(raw_score: f64, selected: &[EvidenceSignal], support_family_count: usize) -> Classification {
    if raw_score <= -10.0 {
        return Classification::Contradictory;
    }

    let has_anchor = selected.iter().any(|s| {
        s.direction == Direction::Support
            && matches!(
                s.signal_type,
                SignalType::EmailExact
                    | SignalType::DirectProfileLink
                    | SignalType::SharedPersonalDomain
                    | SignalType::DomainExact
                    | SignalType::ExactAvatar
                    | SignalType::ProjectOverlap
            )
    });

    let has_direct_link = selected.iter().any(|s| {
        s.direction == Direction::Support
            && s.signal_type == SignalType::DirectProfileLink
            && calculate_contribution(s) >= 35.0
    });

    let material_contradiction = selected.iter().any(|s| {
        s.direction == Direction::Contradict && calculate_contribution(s).abs() >= 15.0
    });

    if raw_score >= 40.0 && support_family_count >= 2 && has_anchor && !material_contradiction {
        Classification::Strong
    } else if raw_score >= 20.0 && !material_contradiction && (support_family_count >= 2 || has_direct_link) {
        Classification::Likely
    } else if raw_score >= 7.0 {
        Classification::Ambiguous
    } else {
        Classification::Weak
    }
}
